lib/engine/game/g_1822/step/route.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/route'
+
+module Engine
+  module Game
+    module G1822
+      module Step
+        class Route < Engine::Step::Route
+          def actions(entity)
+            return [] if !entity.operator? || @game.route_trains(entity).empty? || !@game.can_run_route?(entity)
+
+            @pullman_train ||= nil
+            actions = ACTIONS.dup
+            actions << 'choose' if !@pullman_train && find_pullman_train(entity)
+            actions
+          end
+
+          def attach_pullman
+            @orginal_train = @pullman_train.dup
+            distance = train_city_distance(@pullman_train)
+            @pullman_train.name += '+'
+            @pullman_train.distance = [
+              {
+                'nodes' => %w[city offboard],
+                'pay' => distance,
+                'visit' => distance,
+              },
+              {
+                'nodes' => ['town'],
+                'pay' => 99,
+                'visit' => 99,
+              },
+            ]
+          end
+
+          def choice_name
+            'Choose which train you want to attach your pullman carriage train'
+          end
+
+          def choices
+            choices = {}
+            @game.route_trains(current_entity).each_with_index do |train, index|
+              choices[index.to_s] = "#{train.name} train"
+            end
+            choices
+          end
+
+          def detach_pullman
+            @pullman_train.name = @orginal_train.name
+            @pullman_train.distance = @orginal_train.distance
+
+            @orginal_train = nil
+            @pullman_train = nil
+          end
+
+          def find_pullman_train(entity)
+            entity.trains.find { |t| @game.pullman_train?(t) }
+          end
+
+          def process_choose(action)
+            entity = action.entity
+            @pullman_train = @game.route_trains(entity)[action.choice.to_i]
+            @log << "#{entity.id} chooses to attach the pullman to the #{@pullman_train.name} train"
+
+            attach_pullman
+          end
+
+          def process_run_routes(action)
+            super
+
+            detach_pullman if defined?(@pullman_train) && @pullman_train
```

Comment:
defined? shouldn't be necessary

Reply:
Apart from this looks good to me.

Reply:
No you are correct. I still have some problem with not defining variables as you have to do in a compiling programming language :)

Reply:
As a C++ programmer it doesn't sit well with me either :)